### PR TITLE
Fix missing tick labels on twinned axes.

### LINKED ISSUE #27812: BUG:No xlabel  for twinx axes in subploted figures
#### Code Sample

```python
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import matplotlib as mpl

df=pd.DataFrame()

t1 = np.arange(1, 9.0, 1)
df['TIME'] = t1
df['DAT1'] = t1
df['DAT2'] = np.arange(9, 1.0, -1)

fig = plt.figure(constrained_layout=True)
gs = mpl.gridspec.GridSpec(2, 3, figure=fig)

ax1 = fig.add_subplot(gs[0, :])
ax3 = fig.add_subplot(gs[1, :])

color = 'tab:red'
ax1.set_xlabel('time (s)')
ax1.set_ylabel('exp', color=color)
df.plot(ax=ax1, x='TIME', y='DAT1')
ax1.tick_params(axis='y', labelcolor=color)
**ax2 = ax1.twinx()  # instantiate a second axes that shares the same x-axis

color = 'tab:blue'
ax2.set_ylabel('sin', color=color)  # we already handled the x-label with ax1
df.plot(ax=ax1, x='TIME', y='DAT2')
ax2.tick_params(axis='y', labelcolor=color)**

fig.tight_layout()  # otherwise the right y-label is slightly clipped
plt.show()
```
#### Problem description

When using GridSpec to set the layout, xlabels are not displayed if twinx is used
Figure_1

If the following code is commented out：
“”“
ax2 = ax1.twinx()  # instantiate a second axes that shares the same x-axis

color = 'tab:blue'
ax2.set_ylabel('sin', color=color)  # we already handled the x-label with ax1
df.plot(ax=ax1, x='TIME', y='DAT2')
ax2.tick_params(axis='y', labelcolor=color)
”“”
You can see that xlabel can be displayed：
Figure_2

#### Expected Output
show xlabels
#### Output of ``pd.show_versions()``

<details>

[paste the output of ``pd.show_versions()`` here below this line]
INSTALLED VERSIONS
------------------
commit: None
python: 3.7.3.final.0
python-bits: 64
OS: Windows
OS-release: 10
machine: AMD64
processor: Intel64 Family 6 Model 23 Stepping 10, GenuineIntel
byteorder: little
LC_ALL: None
LANG: zh_CN
LOCALE: None.None

pandas: 0.24.2
pytest: 4.3.1
pip: 19.0.3
setuptools: 40.8.0
Cython: 0.29.6
numpy: 1.16.2
scipy: 1.2.1
pyarrow: None
xarray: None
IPython: 7.4.0
sphinx: 1.8.5
patsy: 0.5.1
dateutil: 2.8.0
pytz: 2018.9
blosc: None
bottleneck: 1.2.1
tables: 3.5.1
numexpr: 2.6.9
feather: None
matplotlib: 3.0.3
openpyxl: 2.6.1
xlrd: 1.2.0
xlwt: 1.3.0
xlsxwriter: 1.1.5
lxml.etree: 4.3.2
bs4: 4.7.1
html5lib: 1.0.1
sqlalchemy: 1.3.1
pymysql: None
psycopg2: None
jinja2: 2.10
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None
gcsfs: None
</details>

### PR DESCRIPTION
**Pull request by ebardie**
## Background:

Multi-row and/or multi-column subplots can utilize shared axes.

An external share happens at axis creation when a sharex or sharey
parameter is specified.

An internal share, or twinning, occurs when an overlayed axis is created
by the `Axes.twinx()` or `Axes.twiny()` calls.

The two types of sharing can be distinguished after the fact in the
following manner. If two axes sharing an axis also have the same
position, they are internally sharing an axis, they are twinned. If their positions are different, they are externally sharing the common axis.

For externally shared axes Pandas automatically removes tick labels for
all but the last row and/or first column in
`./pandas/plotting/_matplotlib/tools.py`'s function `_handle_shared_axes()`.

## The problem:

`_handle_shared_axes()` should be interested in externally shared axes,
whether or not they are also twinned. It should, but **does not**, ignore
axes which are only twinned. 

Which means that twinned-only axes wrongly lose their tick labels.

## The cure:

This commit introduces `_ignore_colocated_axes()` which ignores any
axes in a list which have the same position as the given axis. It uses
this function to envelope calls from `_handle_shared_axes()` to
`Axes.get_shared_{x,y}_axes().get_siblings(ax)`.

## The demonstration test case:

Note especially the axis labels (and associated tick labels).

```python
#!/usr/bin/python3

import matplotlib.pyplot as plt
import numpy as np
import pandas as pd

# Create data
df = pd.DataFrame({'a': np.random.randn(1000),
                   'b': np.random.randn(1000)})

# Create figure
fig = plt.figure()
plots = fig.subplots(2, 3)

# Create *externally* shared axes
plots[0][0] = plt.subplot(231, sharex=plots[1][0])
plots[0][2] = plt.subplot(233, sharex=plots[1][2])

# Create *internally* shared axes
twin_ax1 = plots[0][1].twinx()
twin_ax2 = plots[0][2].twinx()

# Plot data to primary axes
df['a'].plot(ax=plots[0][0], title="External share only").set_xlabel("this label should never be visible")
df['a'].plot(ax=plots[1][0])

df['a'].plot(ax=plots[0][1], title="Internal share (twin) only").set_xlabel("this label should always be visible")
df['a'].plot(ax=plots[1][1])

df['a'].plot(ax=plots[0][2], title="Both").set_xlabel("this label should never be visible")
df['a'].plot(ax=plots[1][2])

# Plot data to twinned axes
df['b'].plot(ax=twin_ax1, color='green')
df['b'].plot(ax=twin_ax2, color='yellow')

# Do it
plt.show()
```
---

- [ ] closes #27812
- [ ] tests added / passed
- [ ] passes `black pandas`
- [ ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

### CODE REVIEW

**Comment by gfyoung** [comment]
@ebardie : Thanks for opening this PR and welcome to `pandas` !

FYI: as a matter of procedure, we generally open issues first before creating the PR to address, just so that we can better assess how to address it in the first place.  No need to close this PR, but just so you know for subsequent contributions.

**Comment by ebardie** [comment]
> @ebardie : Thanks for opening this PR and welcome to `pandas` !
> 
> FYI: as a matter of procedure, we generally open issues first before creating the PR to address, just so that we can better assess how to address it in the first place. No need to close this PR, but just so you know for subsequent contributions.

Thanks. Will do.

File: pandas/plotting/_matplotlib/tools.py [hunk 1]
```diff
@@ -325,7 +357,15 @@ def _handle_shared_axes(axarr, nplots, naxes, nrows, ncols, sharex, sharey):
...
                     continue
-                if sharey or len(ax.get_shared_y_axes().get_siblings(ax)) > 1:
+                if (
+                    sharey
+                    or len(
+                        _ignore_colocated_axes(
+                            ax, ax.get_shared_y_axes().get_siblings(ax)
+                        )
+                    )
+                    > 1
```
Comment (jbrockmendel) [inline pandas/plotting/_matplotlib/tools.py]: this expression and the one above are pretty nasty.  Would it be costly to pre-compute the _ignore_colocated_axes length to make this clearer?

File: pandas/plotting/_matplotlib/tools.py [hunk 2]
```diff
@@ -289,6 +289,22 @@ def _remove_labels_from_axis(axis):
     axis.get_label().set_visible(False)
 
 
+# If two axes sharing an axis also have the same position, they can be referred to as
+# internally sharing the common axis (a.k.a twinning). _handle_shared_axes() is only
+# interested in axes externally sharing an axis, regardless of whether either of the
+# axes is also internally sharing with a third axes.
```
Comment (jbrockmendel) [inline pandas/plotting/_matplotlib/tools.py]: could this go into a docstring for _ignore_colocated_axes?

**Comment by jbrockmendel** [comment]
Can you add a test?

File: pandas/plotting/_matplotlib/tools.py (hunk 1 above)
Comment (ebardie) [reply pandas/plotting/_matplotlib/tools.py]: I've reworked this and it looks much better, thanks.

File: pandas/plotting/_matplotlib/tools.py (hunk 2 above)
Comment (ebardie) [reply pandas/plotting/_matplotlib/tools.py]: Done

**Comment by WillAyd** [comment]
Closing as I think this is stale, but @ebardie certainly ping if you'd like to pick back up and can address comments around test

**Comment by WillAyd** [comment]
Whoops sorry I see test was added; reopening

**Comment by ebardie** [comment]
@WillAyd I presume there's nothing more I need to do other than wait 'til this bubbles to the top of @TomAugspurger's (presumably rather extensive) ToDo list?

**Comment by TomAugspurger** [comment]
I'm not sure about the additional complexity here. If you're using matplotlib to create the axes in the first place, can you not pass your Series / DataFrame to whatever `Axes.plot` your using?

**Comment by jreback** [comment]
can you merge master and will look again

**Comment by WillAyd** [comment]
@ebardie can you merge master?

**Comment by WillAyd** [comment]
Closing as I think stale but ping if you'd like to continue and can merge master

**Comment by ebardie** [comment]
> Closing as I think stale but ping if you'd like to continue and can merge master

I've finally had some time for this! I've rebased on current master. Can you re-open or should I issue a fresh PR?

**Comment by jreback** [comment]
pls open a fresh PR

**Pull request closed without merging**